### ISSUE TITLE
Custom cell processor implementation is confusing.

### ISSUE THREAD

**Issue by robross0606** [super-csv/super-csv#36]
The mechanism for extending CellProcessorAdapter is odd and requires "code by convention" rather than interface enforcement.  Super CSV could learn a thing or two from CSVeed's AbstractConverter code pattern in this particular regard, though just about everything else is better in SuperCSV.  For example, it is entirely unclear how a CellProcessAdapter does bi-directional conversion.  Or is a different adapter required for reading than writing?  If so, why?

**Comment by kbilsted** [comment]
Hi @robross0606 

Thanks for your input. I'm not entirely sure what you mean by lack of interfaces, there is https://github.com/super-csv/super-csv/blob/master/super-csv/src/main/java/org/supercsv/cellprocessor/ift/CellProcessor.java

The reason that a processor should only do one thing, is that e.g. the constructor will take very different input depending on whether it is reading or writing. It would be confusing to get an error when reading and using the constructor meant for writing.

But maybe I'm just misunderstanding you. Can you elaborate with an example or two your thoughts on interfaces and implementations?

**Comment by robross0606** [comment]
Yes, I believe you misunderstanding what I was trying to say.  For starters, “execute” is relatively meaningless. In the context of conversion for CSV, “toString” and “fromString” are far more meaningful interface implementations.  I can’t really tell from the naming of any of this what it actually will do.  Perhaps that’s the point and it is just the documentation (or my knowledge) that is lacking.

 

Secondly, using Object instead of Java Generics leads to even more vagueness in the API.  What the heck is “execute” doing?  Anything?  Everything?  It’s actually TOO generic to be intuitive.  CSVeed’s interface specification is far more obvious, meaningful and easier to implement.  The only complication I see here is the desire to support processor chaining which could complicate the interface.

 

Thirdly, the need to implement both a no-argument constructor and a one argument constructor leads to the possibility of creating an implementation that will satisfy the interface while still failing at runtime.  That is not a desirable API from a development standpoint.

**Comment by kbilsted** [comment]
How can we best improve on the interfaces without breaking backward compatibility? 

It has been many years since I last reviewed the use of generics, but if you have concrete ideas, please bring them forth.

You are more than welcome to come up with a plan that we may agree upon, and then a PR for merging.

**Comment by jamesbassett** [comment]
I agree that custom cell processors are a bit quirky - but I think the [documentation](http://super-csv.github.io/super-csv/examples_new_cell_processor.html) at least points out these quirks.  If you're familiar with servlet filters in JEE, it's actually a very similar pattern (without the chaining of constructors and the use of interfaces to restrict how they're chained together).

If you look at the source, you'll see that cell processors are not tied to reading or writing (they can be used for both), and some don't do any transformation (they are just for validation, or capturing data). So `execute()`, while extremely vague,  is probably not that terrible (see `doFilter()` in JEE servlet filters).

When I first came on board I was keen to add generics to cell processors - until I saw how this doesn't work because of the chaining of cell processors.

Without some kind of heart surgery and breaking backwards compatibility I don't think there's a lot we can do, apart from expand upon the documentation.
